### PR TITLE
14729 move testrulebutton to toolbar

### LINKED ISSUE #14729: Alert Tab UX Update Step2
Continuation on https://github.com/grafana/grafana/issues/14388 

* Move AlertTab & StateHistory from dashgrid to app/features/alerting

* Move test rule to react and add it as toolbar button (like QueryInspector in Queries tab)

### PR DESCRIPTION
**Pull request by hugohaggmark**
Fixes #14729 
- Moved TestRule button to toolbar
- Fixed small bug when toggling items in toolbar

### CODE REVIEW

File: public/app/features/alerting/TestRuleButton.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  render() {
+    const { testRuleResponse, isLoading } = this.state;
+    const { LoadingPlaceholder } = this.props;
+
+    if (isLoading === true) {
+      return <LoadingPlaceholder text="Evaluating rule" />;
+    }
+
+    return (
+      <>
```
Comment (dprokop) [inline public/app/features/alerting/TestRuleButton.tsx]: There's need for the fragment here

File: public/app/features/dashboard/dashgrid/EditorTabBody.tsx [hunk 2]
```diff
@@ -52,7 +52,7 @@ export class EditorTabBody extends PureComponent<Props, State> {
   onToggleToolBarView = (item: EditorToolbarView) => {
     this.setState({
       openView: item,
-      isOpen: !this.state.isOpen,
+      isOpen: this.state.openView !== item ? true : !this.state.isOpen,
```
Comment (dprokop) [inline public/app/features/dashboard/dashgrid/EditorTabBody.tsx]: You can simplify this line like that:
```ts
isOpen: this.state.openView !== item || !this.state.isOpen
```

File: public/app/features/alerting/TestRuleButton.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  LoadingPlaceholder: any;
+}
+
+interface State {
+  isLoading: boolean;
+  testRuleResponse: {};
+}
+
+export class TestRuleButton extends PureComponent<Props, State> {
+  constructor(props) {
```
Comment (dprokop) [inline public/app/features/alerting/TestRuleButton.tsx]: No need for this constructor, as initial state does not rely on the props. See https://github.com/grafana/grafana/pull/14763#discussion_r246006729 for details

File: public/app/features/alerting/TestRuleButton.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import React, { PureComponent } from 'react';
+import { JSONFormatter } from 'app/core/components/JSONFormatter/JSONFormatter';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+import { DashboardModel } from '../dashboard/dashboard_model';
+
+export interface Props {
+  panelId: number;
+  dashboard: DashboardModel;
+  LoadingPlaceholder: any;
```
Comment (dprokop) [inline public/app/features/alerting/TestRuleButton.tsx]: Would be awesome if we could create a generic interface for handling `LoadingPlaceholder` prop down the road. Extracting `LoadingPlaceholder` from `AlertTab` component file and placing it in grafana/ui package would be a good starting point.

Also, having prop names in camelCase would be good(vide [Airbnb styleguide](https://github.com/airbnb/javascript/tree/master/react#props))

File: public/app/features/alerting/TestRuleButton.tsx (hunk 3 above)
Comment (hugohaggmark) [reply public/app/features/alerting/TestRuleButton.tsx]: Thanks, will fix!

File: public/app/features/dashboard/dashgrid/EditorTabBody.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/dashboard/dashgrid/EditorTabBody.tsx]: Thanks, will fix!

File: public/app/features/alerting/TestRuleButton.tsx (hunk 1 above)
Comment (hugohaggmark) [reply public/app/features/alerting/TestRuleButton.tsx]: Good catch

File: public/app/features/alerting/TestRuleButton.tsx (hunk 4 above)
Comment (hugohaggmark) [reply public/app/features/alerting/TestRuleButton.tsx]: Good suggestion, copied existing LoadingPlaceHolder from QueryInspector. Not sure if this should be part of this PR though.

Comment (dprokop) [reply public/app/features/alerting/TestRuleButton.tsx]: Good point. Created an issue for that #14793

**Review by Ijin08 — CHANGES_REQUESTED** [review]
Looks good! The only issue I had was if I have Test Rule open and delete the alert it stays open.

screenshot 2019-01-09 at 16 38 23

**Comment by torkelo** [comment]
Not sure the test rule showing after it's deleted is super critical (and is not super easy to fix right now).

**Comment by torkelo** [comment]
Great job!

**Comment by hugohaggmark** [comment]
@Ijin08 yes I saw this as well, the same issue was already there for the State History button though. If you click on that first and then delete you have the same behaviour.

**Pull request merged**